File: Lectures/Lecture 7/python/exercise_binary_tree.py

# Node beskriver ett binärt träd
class Node:
    def __init__(self, name):
        self.name = name
        self.left = None
        self.right = None

    def size(self):
        s = 1

        if self.left is not None:
            s += self.left.size()
        if self.right is not None:
            s += self.right.size()

        return s

    def height(self, height=1):
        # Valfritt: Explicit basfall. Hanteras av koden nedanför om vi skippar detta.
        if self.left is None and self.right is None:
            return height

        # Vilken höjd får vi tillbaka från våra barn
        left_height = height
        right_height = height

        if self.left is not None:
            left_height = self.left.height(height + 1)
        if self.right is not None:
            right_height = self.right.height(height + 1)

        # Vilken av höjderna är störst
        return max(left_height, right_height)

    def level(self, root, distance=0):
        # Self är vår nod nere i träder och root är noden vi vill jämföra med.
        # Rekursera och gå vänster/höger genom att byta ut root.
        # DVS: Vi börjar på en nod och vandrar runt tills vi hittar "self"-noden

        # Har vi hittat rätt nod?
        if self == root:
            return distance

        if root.left is not None:
            # Rekursivt anrop åt vänster om det finns något där
            left_level = self.level(root.left, distance + 1)
            if left_level is not None:
                # Om sökningen lyckades, vi hittade self-noden genom att gå åt vänster
                return left_level

        if root.right is not None:
            # Rekursivt anrop åt höger om det finns något där
            right_level = self.level(root.right, distance + 1)
            if right_level is not None:
                # Om sökningen lyckades, vi hittade self-noden genom att gå åt höger
                return right_level

        # Om noden vi besöker inte är self-noden vi letar efter,
        # men vi inte kan gå åt något håll (vänster eller höger),
        # så måste vi indikera att vi misslyckades. Därför skickar
        # vi tillbaka "None"
        return None


if __name__ == '__main__':
    root_node = Node('uggla')
    n1 = Node('hamster')
    n2 = Node('apa')
    n3 = Node('gorilla')
    n4 = Node('häst')
    n5 = Node('lemur')
    n6 = Node('pingvin')
    n7 = Node('varg')

    root_node.left = n1
    root_node.right = n6
    n1.left = n2
    n1.right = n4
    n2.right = n3
    n4.right = n5
    n6.right = n7

    print(root_node.size())
    print(n6.size())
    print(n7.size())

    print()
    print(root_node.height())

    print()
    print(n7.level(root_node))
    print(n3.level(n5))

File: Lectures/Lecture 7/cpp/exercise_binary_tree.cpp
#include <string>
#include <iostream>

using namespace std;

/** Node beskriver ett binärt träd (rekursivt)
 * 
*/
class Node {
public:
    string name;
    Node * left;
    Node * right;

    Node(string name) {
        this->name = name;
        this->left = nullptr;
        this->right = nullptr;
    }

    int size() {
        int s = 1;

        if (this->left != nullptr) {
            s += this->left->size();
        }
        if (this->right != nullptr) {
            s += this->right->size();
        }

        return s;
    }

    int height(int height = 1) {
        // Vilken höjd får vi från vara barn?
        // Om vi inte har vänster eller höger barn, så blir nuvarande höjd max höjden
        int left_height = height;
        int right_height = height;

        if (this->left != nullptr) {
            left_height = this->left->height(height + 1);
        }
        if (this->right != nullptr) {
            right_height = this->right->height(height + 1);
        }

        return (left_height > right_height ? left_height : right_height);
    }

    int level(Node* root, int distance = 0) {
        // Vi vill gå rekursivt tills vi hittar vår nod, och räkna antal nivåer mellan

        if (this == root) {
            return distance;
        }

        if (root->left == nullptr && root->right == nullptr) {
            // Vi har kommit till botten av trädet, men ändå inte hittat vår nod!
            // Returnera -1 för att indikera att vi inte hittat den
            return -1;
        }

        int find_distance = -1;

        if (root->left != nullptr) {
            find_distance = level(root->left, distance+1);
            if (find_distance != -1) {
                return find_distance;
            }
        }
        if (root->right != nullptr) {
            find_distance = level(root->right, distance+1);
            if (find_distance != -1) {
                return find_distance;
            }
        }

        // Vi hittade den inte!
        return -1;
    }
};

int main() {
    Node root_node = Node("uggla");
    Node n1 = Node("hamster");
    Node n2 = Node("apa");
    Node n3 = Node("gorilla");
    Node n4 = Node("häst");
    Node n5 = Node("lemur");
    Node n6 = Node("pingvin");
    Node n7 = Node("varg");

    root_node.left = &n1;
    root_node.right = &n6;
    n1.left = &n2;
    n1.right = &n4;
    n2.right = &n3;
    // n3 has no children
    n4.right = &n5;
    // n5 has no chilcren
    n6.right = &n7;
    // n7 has no children
    
    cout << "root size:" << root_node.size() << endl;
    cout << "n6 size:" << n6.size() << endl;
    cout << "n7 size:" << n7.size() << endl;

    cout << endl;
    cout << "root height:" << root_node.height() << endl;
    
    cout << endl;
    cout << "n7 level:" << n7.level(&root_node) << endl;
    cout << "n3 level against n5:" << n3.level(&n5) << endl;
}